### PR TITLE
visionlan gpu updates

### PR DESCRIPTION
**Pull request by wtomin**
Thank you for your contribution to the MindOCR repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

Visionlan training (on Ascend, amp_level=O2), evaluation, and readme, docs.

### CODE REVIEW

File: configs/rec/visionlan/README_CN.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+## 3.快速入门
+
+### 3.1安装
+
+请参考[MindOCR中的安装说明]((https://github.com/mindspore-lab/mindocr#installation))。
+
+### 3.2数据集准备
+
+* 训练集
+VisionLAN的作者使用了两个合成文本数据集进行训练：SynthText（800k）和MJSynth。请按照[原始VisionLAN repository](https://github.com/wangyuxin87/VisionLAN)的说明进行操作，或从[BaiduYun](https://pan.baidu.com/s/1_2dqqxW1vDL9t3B-jlAYRw)（密码：z0r5）下载这两个LMDB数据集，或从[Openi Platform](https://openi.pcl.ac.cn/ddeng/ocr_datasets_visionlan/datasets)下载这两个LMDB数据集。
```
Comment (zhtmike) [inline configs/rec/visionlan/README_CN.md:83]: 尽量用官方源作为链接，避免用个人仓。另外如果数据集和官方不一致的话，请提供如何产生LMDB文件的方法，类似 https://github.com/mindspore-lab/mindocr/blob/0a7d69da0d16ed305d14b004ec0a92ae914d0673/configs/rec/master/README_CN.md#3122-synthtext-dataset

File: configs/rec/visionlan/README_CN.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+datasets
+├── evaluation
+│   ├── Sumof6benchmarks
+│   ├── CUTE
+│   ├── IC13
+│   ├── IC15
+│   ├── IIIT5K
+│   ├── SVT
+│   └── SVTP
+└── train
```
Comment (zhtmike) [inline configs/rec/visionlan/README_CN.md:108]: Synthtext有很多版本，请描述下每个dataset 数量，参考crnn readme

File: mindocr/data/rec_lmdb_dataset.py [hunk 3]
```diff
@@ -189,9 +190,12 @@ def __getitem__(self, idx):
         data = {"img_lmdb": sample_info[0], "label": sample_info[1]}
 
         # perform transformation on data
-        data = run_transforms(data, transforms=self.transforms)
-        output_tuple = tuple(data[k] for k in self.output_columns)
-
+        try:
```
Comment (zhtmike) [inline mindocr/data/rec_lmdb_dataset.py:193]: 这个修改要确保只能在训练集有效，验证集可能会因为这个导致精度误差。

File: mindocr/losses/rec_loss.py [hunk 4]
```diff
@@ -57,6 +57,82 @@ def construct(self, pred: Tensor, label: Tensor) -> Tensor:
         return loss
 
 
+class VisionLANLoss(LossBase):
+    def __init__(self, mode="LF_1", weight_res=0.5, weight_mas=0.5, reduction="mean", **kwargs):
+        super().__init__()
+        self.criterion = nn.CrossEntropyLoss(
```
Comment (zhtmike) [inline mindocr/losses/rec_loss.py:63]: ignore_index 应该是指index，而不是value，最好确认下

File: configs/rec/visionlan/README_CN.md (hunk 1 above)
Comment (wtomin) [reply configs/rec/visionlan/README_CN.md:83]: 官方源提供的链接就是百度云。但是百度云下载起来很慢。我从官方提供的百度云下载下来以后，又上传到openi平台上我自己建的一个repo下面了。
我可以把百度云和openi的链接删除。用户从原repo的说明就能找到百度云下载链接。

Comment (wtomin) [reply configs/rec/visionlan/README_CN.md:83]: 原作者提供的下载链接就是百度云。我也不确定作者提供的synthtext和MJsynth lmdb 文件是不是跟官方网站的一致。
保险起见，我把baiduyun和openi链接删掉，只引导用户去原repo寻找下载方法。

File: configs/rec/visionlan/README.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+</div>
+
+<details open markdown>
+  <div align="center">
+  <summary>Detailed accuracy results for six benchmark datasets</summary>
+
+  | **Model** |  **Context** | **IIIT5k_3000** | **IC13_857** |  **SVT** | **IC15_1811** | **SVTP** | **CUTE80** | **Average** |
+  | :------: | :------: | :------: | :------: | :------: | :------: | :------: | :------: |:------: |
+  | visionlan |  GTX3090x8-MS2.0-G |  0.9537  |  0.9557  |  0.9289  | 0.8283  |  0.8450  |  0.8958|  0.9012|
```
[the remark this answers is not in the document]
Comment (wtomin) [reply configs/rec/visionlan/README.md]: 已经修改为10个数据集的平均精度

File: mindocr/data/rec_lmdb_dataset.py (hunk 3 above)
Comment (wtomin) [reply mindocr/data/rec_lmdb_dataset.py]: 这个是之前debug时加的一个exception，我注意到实际训练时从来没有print exception message, 可见这段可以删除。已删。

File: configs/rec/visionlan/README_CN.md (hunk 2 above)
Comment (wtomin) [reply configs/rec/visionlan/README_CN.md:153]: 增加数据集大小描述

File: mindocr/losses/rec_loss.py (hunk 4 above)
Comment (wtomin) [reply mindocr/losses/rec_loss.py:63]: 确定, ignore_index指的是填充值：https://www.mindspore.cn/docs/zh-CN/master/api_python/nn/mindspore.nn.CrossEntropyLoss.html

File: configs/rec/visionlan/README.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
+English | [中文](README_CN.md)
+
+# VisionLAN
+
+<!--- Guideline: use url linked to abstract in ArXiv instead of PDF for fast loading.  -->
+
+> VisionLAN: [From Two to One: A New Scene Text Recognizer with Visual Language Modeling Network](https://arxiv.org/abs/2108.09661)
```
Comment (zhtmike) [inline configs/rec/visionlan/README.md:7]: 不需要outline，格式跟CRNN对齐就行

File: mindocr/data/transforms/rec_transforms.py [hunk 7]
```diff
@@ -133,6 +135,77 @@ def __call__(self, data: dict):
         return data
 
 
+class VisionLANLabelEncode(RecCTCLabelEncode):
+    """Encode the VisionLAN labels"""
+
+    def __init__(
+        self, max_text_len, character_dict_path=None, use_space_char=False, blank_at_last=True, lower=False, **kwargs
```
Comment (zhtmike) [inline mindocr/data/transforms/rec_transforms.py:142]: 加一下docstring

File: mindocr/losses/rec_loss.py [hunk 8]
```diff
@@ -68,4 +144,4 @@ def construct(self, logits: Tensor, labels: Tensor) -> Tensor:
         num_classes = logits.shape[-1]
         logits = ops.reshape(logits, (-1, num_classes))
         labels = ops.reshape(labels, (-1,))
-        return self.criterion(logits, labels)
+        return self.criterion(logits, labels)
```
Comment (zhtmike) [inline mindocr/losses/rec_loss.py:147]: 少了最后\n

**Review by zhtmike — APPROVED** [review]

File: tools/benchmarking/multi_dataset_eval.py [hunk 9]
```diff
@@ -201,11 +204,11 @@ def parse_args():
 
 
 if __name__ == "__main__":
-    # argpaser
-    args = parse_args()
-    yaml_fp = args.config
-    with open(yaml_fp) as fp:
-        config = yaml.safe_load(fp)
+    # # argpaser
+    # args = parse_args()
```
Comment (zhtmike) [inline tools/benchmarking/multi_dataset_eval.py:208]: 请清一下commented掉的code

File: configs/rec/visionlan/README_CN.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+### 3.2数据集准备
+
+* 训练集
+
+VisionLAN的作者使用了两个合成文本数据集进行训练：SynthText（800k）和MJSynth。请按照[原始VisionLAN repository](https://github.com/wangyuxin87/VisionLAN)的说明进行操作，下载这两个LMDB数据集。
+
+下载`SynthText.zip`和`MJSynth.zip`后，请解压缩并将它们放置在`./datasets/train`目录下。训练集数据总共包括 14,200,701 个样本。更多关于训练集的信息如下:
+
+```text
+- [SynText](http://www.robots.ox.ac.uk/~vgg/data/scenetext/): 25GB, 6,976,115 样本
```
Comment (HaoyangLee) [inline configs/rec/visionlan/README_CN.md:77]: markdown显示格式有误

File: mindocr/models/backbones/rec_resnet45.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        x = self.layer5(x)
+        return x
+
+    def construct(self, x: Tensor) -> Tensor:
+        x = self.forward_features(x)
+        return [x]
+
+
+@register_backbone
+def rec_resnet45(pretrained: bool = False, strides: List = [2, 2, 2, 1, 1], in_channels: int = 3, **kwargs):
```
Comment (HaoyangLee) [inline mindocr/models/backbones/rec_resnet45.py:242]: rec_resnet45是否可以在rec_resnet.py中register_backbone，而不单独创建rec_resnet45.py？两者结构差别很大？

**Comment by HaoyangLee** [comment]
mindocr/models目录下，没有rec_visionlan.py，请参考其他模型。

File: mindocr/models/backbones/rec_resnet45.py (hunk 11 above)
Comment (wtomin) [reply mindocr/models/backbones/rec_resnet45.py:242]: 结构不一样。rec_resnet.py里的RecResNet 是由三层conv layer + maxpool + block1 + block2 + block3 + block4 + maxpool组成的。rec_resnet45.py里的RecResNet45是由一层conv layer + block1 + block2 + block3+ block4 + block5 组成的。我认为需要单独创建rec_resnet45.py.

**Comment by wtomin** [comment]
rec_visionlan已经更新，build_model(*visionlan_resnet45", pretrained=True) 没问题

File: configs/rec/visionlan/README.md [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+
+## 1. Introduction
+
+### 1.1 VisionLAN
+
+ Visual Language Modeling Network (VisionLAN) [<a href="#references">1</a>] is a text recognion model that learns the visual and linguistic information simultaneously via **character-wise occluded feature maps** in the training stage. This model does not require an extra language model to extract linguistic information, since the visual and linguistic information can be learned as a union.
```
Comment (Songyuanwei) [inline configs/rec/visionlan/README.md:13]: [<a href="#references">1</a>] 这个和下面写的不一样，没有链接。

File: tools/benchmarking/multi_dataset_eval.py [hunk 13]
```diff
@@ -121,7 +124,8 @@ def main(cfg):
     for dirpath, dirnames, _ in os.walk(data_dir_root + "/"):
         if not dirnames:
             dataset_config = copy.deepcopy(cfg.eval.dataset)
-            dataset_config["data_dir"] = dirpath
+            dataset_config["data_dir"] = dirpath.replace(data_dir_root + "/", "")
```
Comment (zhtmike) [inline tools/benchmarking/multi_dataset_eval.py:127]: 这个修改作用是什么？

File: mindocr/postprocess/rec_postprocess.py [hunk 14]
```diff
@@ -156,6 +156,71 @@ def __call__(self, preds: Union[Tensor, np.ndarray], labels=None, **kwargs):
         return {"texts": texts, "confs": confs, "raw_chars": raw_chars}
 
 
+class VisionLANPostProcess(RecCTCLabelDecode):
+    """Convert between text-label and text-index"""
+
```
Comment (zhtmike) [inline mindocr/postprocess/rec_postprocess.py:161]: 加一下docstring~

File: mindocr/postprocess/rec_postprocess.py [hunk 15]
```diff
@@ -156,6 +156,71 @@ def __call__(self, preds: Union[Tensor, np.ndarray], labels=None, **kwargs):
...
+    def __init__(self, character_dict_path=None, use_space_char=False, blank_at_last=True, lower=False, **kwargs):
+        super(VisionLANPostProcess, self).__init__(character_dict_path, use_space_char, blank_at_last, lower)
+        self.max_text_length = kwargs.get("max_text_length", 25)
```
Comment (zhtmike) [inline mindocr/postprocess/rec_postprocess.py:164]: max_text_length 这搬到argument里面，用户容易看漏

File: tools/benchmarking/multi_dataset_eval.py (hunk 13 above)
Comment (wtomin) [reply tools/benchmarking/multi_dataset_eval.py:127]: 因为我的测试集路径出错了。

我想要测试"./datasets/test"下所有的数据集，假设这个文件夹下面有"ic15"。 给multi_dataset_eval.py传入--eval.dataset.dataset_root="./datasets"和--eval.dataset.data_dir="test", 然后照源代码跑，就发现程序会去load “./datasets/./datasets/test/ic15”， 从而报错。

Comment (zhtmike) [reply tools/benchmarking/multi_dataset_eval.py:127]: 可以考虑换成 dataset_config["data_dir"] = os.path.abs(dirpath)， 自动规范输入路径

File: configs/rec/visionlan/visionlan_resnet45_LA.yaml [hunk 16]
```diff
@@ -0,0 +1,172 @@
+system:
+  mode: 0 # 0 for graph mode, 1 for pynative mode in MindSpore
+  distribute: True
+  amp_level: 'O2'
+  seed: 42
+  log_interval: 200
+  val_while_train: True
+  drop_overflow_update: False
+  training_step: &training_step 'LA'
```
Comment (hqkate) [inline configs/rec/visionlan/visionlan_resnet45_LA.yaml:9]: `training_step` 这个参数感觉是visionlan才用到的，是否可以考虑放到 common 底下

File: mindocr/postprocess/rec_postprocess.py (hunk 15 above)
Comment (wtomin) [reply mindocr/postprocess/rec_postprocess.py]: 解决了

File: configs/rec/visionlan/visionlan_resnet45_LA.yaml (hunk 16 above)
Comment (wtomin) [reply configs/rec/visionlan/visionlan_resnet45_LA.yaml]: OK

File: mindocr/postprocess/rec_postprocess.py (hunk 14 above)
Comment (wtomin) [reply mindocr/postprocess/rec_postprocess.py:179]: 好的

File: tools/benchmarking/multi_dataset_eval.py (hunk 13 above)
Comment (wtomin) [reply tools/benchmarking/multi_dataset_eval.py]: 可以

**Review by HaoyangLee — APPROVED** [review]
